editors/code/src/main.ts

```diff
@@ -8,15 +8,76 @@ import { activateInlayHints } from './inlay_hints';
 import { activateStatusDisplay } from './status_display';
 import { Ctx } from './ctx';
 import { Config, NIGHTLY_TAG } from './config';
-import { log, assert } from './util';
+import { log, assert, nearestParentWithCargoToml, createWorkspaceWithNewLocation } from './util';
 import { PersistentState } from './persistent_state';
 import { fetchRelease, download } from './net';
 import { spawnSync } from 'child_process';
 import { activateTaskProvider } from './tasks';
 
 let ctx: Ctx | undefined;
+const foundProjects: Set<string> = new Set();
```

Comment:
As a general comment, global singletons should not be added without very good reason.

Reply:
100% agree,
I only used it here because it's not meant be shared outside main, and it didn't fit in Ctx since this is meant to live after a restart.